### ISSUE TITLE
Entering invalid value to DatePicker does not show DatePicker as invalid in browser

### ISSUE THREAD

**Issue by tepi** [vaadin/vaadin-date-picker-flow#223]
To reproduce the issue:

1. Create a new DatePicker
2. Enter "jksghk" and blur the DatePicker
3. DatePicker looks valid

Additionally, there is no easy way to validate invalid input from Java side and set the DatePicker into invalid mode.

**Comment by yuriy-fix** [comment]
Requires complex changes both on the components and flow counterparts logic as well as to the flow Binder behavior: https://github.com/vaadin/components-team-tasks/issues/489

**Comment by alvarezguille** [comment]
More info in epic in https://github.com/vaadin/components-team-tasks/issues/528

**Comment by tomivirkki** [comment]
Could be fixed by synchronizing the `_inputValue` property on event X and then setting the validity server side based on the property value and `value` of the component; The component should be invalid when the `value` is null but there is an input value. Should also validate on blur (and run the Binder validation, same as Value change?).

This isn't related to Binder necessarily. Like other component validators, this will not work with Binder before `getDefaultValidator`  API is implemented (another issue).

**Comment by pekam** [comment]
The validation at Java side has so far been a function of the server-side value, executed on `ValueChangeEvent`. To resolve this issue, it should be a function of the server-side value AND the input's string-value (web component's `_inputValue`).

There is no 1-to-1 relationship between the server-side value and `_inputValue`: value `null` can correspond to `_inputValue=""` or `_inputValue="asdf"` (or any other string that can't be resolved to a date). This means, that it's not enough to run the validation on `ValueChangeEvent`: When the `_inputValue` changes between empty string and invalid date string, the validation status should be updated, but `ValueChangeEvent` is not fired, because the value remains `null` during these changes.

So there are two problems to solve:
- `_inputValue` should be synchronized to server-side
- Validation should also run on `blur` events

Also, we should keep in mind that similar fix migh be needed for other components (at least `IntegerField` and `BigDecimalField`, where non-parseable numbers are converted to `null` server-side value).

### Syncing the `_inputValue`

We don't want to synchronize on the text-field's `value-changed-event`, because that would create network traffic for every keypress.
The two cases when we want to update the input value are the same as when we want to validate: date picker's `value-changed` and `blur`:

datepickerConnector.js:
```javascript
datepicker.addEventListener('value-changed',
    e => setTimeout(() => datepicker.dispatchEvent(new CustomEvent('input-value-changed'))));
datepicker.addEventListener('blur',
    e => setTimeout(() => datepicker.dispatchEvent(new CustomEvent('input-value-changed'))));
// (setTimeout is needed to get the correct string value due to wc implementation)
```
DatePicker.java:
```java
@Synchronize(property = "_inputValue", value = "input-value-changed")
private String getInputValue() {
    return getElement().getProperty("_inputValue", "");
}
```

### Running validation on blur

On component level it's trivial:
```java
addBlurListener(e -> validate());
```
The problem is when you use this together with Binder validators.

Once we implement `HasValidator::getDefaultValidator`, Binder will take also this new `_inputValue` validation into account, but Binder is still validating only on `ValueChangeEvent`s.

So there would still be a bug in this scenario:

1. datepicker is configured as required through Binder
2. input value is changed to "asdf" and field is blurred -> value is null and the field is invalid (via component's blur listener and internal validation)
3. input value is changed to "" and field is blurred -> only component-level validation runs as there's no value-change to trigger Binder validation, field becomes valid although the field was set required through Binder

The logical solution would be, that if the component validates on blur events, Binder should as well.

**Proposal to Flow**: Change Binder to also validate on blur if field implements `BlurNotifier`.
- This would also introduce the validate-after-first-visit -behavior that we have on our web components to Flow users (eg. when you visit an empty field without changing the value)
  - Probably a good thing, since that's how our components have been designed to work?
  - Behavior is changing - can it be done in a patch release? Is it a breaking change?
- Possibly a problem for components implementing `HasValueChangeMode`: With value-change-modes `LAZY` and `TIMEOUT`, the value is synchronized after a delay, so at the time of the blur event, server-side may have the old value.
  - As described [here](https://github.com/vaadin/vaadin-text-field-flow/issues/270), it might be a good idea to sync value on blur also with lazy and timeout modes for other reasons, which would also solve this problem.
 
**Another way to approach this** would be to somehow notify from the component to the Binder that validators should run.
- Eg. `HasValidator::addValidationListener`
- Or with a different kind of approach: `HasValidator::getEventsToValidate`
- It could also be in eg. `HasValue` and have a default implementation for notifying only on `ValueChangeEvent`.
- As a downside, this approach would expose public API that shouldn't really be visible to users, just because it needs to be visible to Binder (similarly to the existing `HasValidator::getDefaultValidator` method).
- Another wild idea, to avoid adding public methods: optional annotation on the component-class, eg. `@ValidationEvents({ValueChangeEvent.class, BlurEvent.class})`

**Comment by pekam** [comment]
After releasing a web component fix that blocks this, I'll open a fixing PR for DatePicker (adding the logic to component-level validation).

After this change, the component should validate correctly when NOT using Binder.

To validate correctly in all cases including Binder, two more steps are needed:
1. Binder should validate on blur: https://github.com/vaadin/flow/issues/8242
  (this one has a simple workaround: `datePicker.addBlurListener(e -> binder.validate())`
2. DatePicker should implement `getDefaultValidator()` (requires bumping the minor version)

If/when the upcoming DatePicker PR is released, but neither of the above-mentioned steps are, the following problems will occur when using Binder:
- Whenever the value changes, only Binder validation is effective and component validation dismissed
  - this is an old issue, but it's important to note that the new fix is dismissed as well, when you're changing from a non-null value to random input like "asdf" (of course it's still marked as invalid if Binder doesn't allow `null` values)
- Focusing and blurring the field without changing the value will trigger only the component's internal validation, dismissing Binder validation (REGRESSION)

If/when step 1 is implemented and step 2 not, the following problem will occur when using Binder:
- This new new fix (invalidate-unparseable-date-string) won't work at all, similarly to the component's validation constraints
  - Binder will always validate after the component has validated, but the Binder doesn't take component's validation implementation into account due to missing step 2.

To avoid introducing a regression, and to make this fix and the existing component-level validation work with Binder, I'd vote for releasing all of these changes in the same platform version. Since step 2 needs a minor bump, this would mean postponing the fix until Vaadin 14.3.

**Comment by pekam** [comment]
Proposal https://github.com/vaadin/flow/issues/8242 is not accepted by Flow team. There will be probably instead some kind of `addValidationStatusListener` in the `HasValidator` interface to allow components to notify Binder about re-validation. The fix to this issue is blocked by the Flow enhancement.

Here's a branch with the changes excluding the implementation of the not-yet-existing event: https://github.com/vaadin/vaadin-date-picker-flow/compare/invalid-date-string-fix?expand=1
